rewrite-maven/src/main/java/org/openrewrite/maven/AddManagedDependency.java

```diff
@@ -0,0 +1,167 @@
+                    "is used when adding a runtime, provided, or import dependency.",
+            example = "import",
+            valid = {"import", "runtime", "provided", "test"},
+            required = false)
+    @Nullable
+    private final String scope;
+
+    @Option(displayName = "Version pattern",
+            description = "Allows version selection to be extended beyond the original Node Semver semantics. So for example," +
+                    "Setting 'version' to \"25-29\" can be paired with a metadata pattern of \"-jre\" to select 29.0-jre",
+            example = "-jre",
+            required = false)
+    @Nullable
+    private String versionPattern;
+
+    @Option(displayName = "Type",
+            description = "The type of dependency to add. If omitted Maven defaults to assuming the type is \"jar\".",
+            valid = {"jar", "pom", "war"},
+            example = "pom",
+            required = false)
+    @Nullable
+    private String type;
+
+    @Option(displayName = "Releases only",
+            description = "Whether to exclude snapshots from consideration when using a semver selector",
+            example = "true",
+            required = false)
+    @Nullable
+    private Boolean releasesOnly;
+
+    @Option(displayName = "Only if using glob pattern",
+            description = "Used to determine if the dependency will be added and in which scope it should be placed.",
+            example = "org.apache.logging.log4j.*")
+    private String onlyIfUsing;
+
+    @Option(displayName = "Classifier",
+            description = "A Maven classifier to add. Most commonly used to select shaded or test variants of a library",
+            example = "test",
+            required = false)
+    @Nullable
+    private String classifier;
+
+    @Option(displayName = "Add to the root pom",
+            description = "Add to the root pom where root is the eldest parent of the pom within the source set.",
+            example = "true",
+            required = false)
+    private Boolean addToRootPom;
+
+    @Override
+    public Validated validate() {
+        Validated validated = super.validate();
+        //noinspection ConstantConditions
+        if (version != null) {
+            validated = validated.or(Semver.validate(version, versionPattern));
+        }
+        return validated;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Add Managed Maven Dependency";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new UsesType<>(onlyIfUsing);
+    }
+
+    @Override
+    protected List<SourceFile> visit(List<SourceFile> before, ExecutionContext ctx) {
+        Map<JavaProject, Set<String>> sourceSetsByProject = new HashMap<>();
+        final AtomicReference<SourceFile> rootPom = new AtomicReference<>();
```

Comment:
There can be more than one "root" project. 

for example, a pom at the top level could have 32 sub-modules. 

However, each sub-module could have a different parent.

Reply:
I see thanks for pointing this out